packages/loader/container-loader/src/container.ts

```diff
@@ -631,13 +631,13 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         this.connectionStateHandler = createConnectionStateHandler(
             {
                 logger: this.mc.logger,
-                connectionStateChanged: (value, oldState, reason) => {
+                connectionStateChanged: (value, oldState, disconnectedReason) => {
                     if (value === ConnectionState.Connected) {
                         this._clientId = this.connectionStateHandler.pendingClientId;
                     }
-                    this.logConnectionStateChangeTelemetry(value, oldState, reason);
+                    this.logConnectionStateChangeTelemetry(value, oldState, disconnectedReason);
                     if (this._lifecycleState === "loaded") {
-                        this.propagateConnectionState(false /* initial transition */);
+                        this.propagateConnectionState(false /* initial transition */, disconnectedReason);
```

Comment:
Ok let's have this be the transition point from generic `reason` to `disconnectedReason`: Do `value === ConnectionState.Disconnected ? reason : undefined`.

This way the "caught up" reason you pointed out can flow to `logConnectionStateChangeTelemetry` as-is but the actual event API can be scoped to only disconnect reason (trying to minimize change to API surface - if it were generic reason then I'd argue we should also update "connected" event)

Reply:
@markfields sounds good, I like this better too. Updated it.